rust/src/optimize.rs

```diff
@@ -0,0 +1,356 @@
+        size: DeltaDataTypeLong,
+    ) -> Result<Action, DeltaTableError> {
+        let deletion_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let deletion_time = deletion_time.as_millis() as i64;
+
+        Ok(Action::remove(action::Remove {
+            path: path.to_string(),
+            deletion_timestamp: Some(deletion_time),
+            data_change: false,
+            extended_file_metadata: None,
+            partition_values: Some(partitions.to_owned()),
+            size: Some(size),
+            tags: None,
+        }))
+    }
+
+    pub async fn execute(mut self, table: &mut DeltaTable) -> Result<Metrics, DeltaWriterError> {
+        //Read files into memory and write into memory. Once a file is complete write to underlying storage.
+        let mut actions = vec![];
+
+        for (_partition_path, merge_partition) in self.operations.iter() {
+            let partition_values = &merge_partition.partition_values;
+            let bins = &merge_partition.bins;
+            debug!("{:?}", bins);
+            debug!("{:?}", _partition_path);
+
+            for bin in bins {
+                let mut writer = RecordBatchWriter::for_table(table, HashMap::new())?;
+
+                for path in &bin.files {
+                    //Read the file into memory and append it to the writer.
+
+                    let parquet_uri = table.storage.join_path(&table.table_uri, path);
+                    let data = table.storage.get_obj(&parquet_uri).await?;
+                    let size: DeltaDataTypeLong = data.len().try_into().unwrap();
+                    let data = SliceableCursor::new(data);
+                    let reader = SerializedFileReader::new(data)?;
+                    let records = reader.metadata().file_metadata().num_rows();
+
+                    let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(reader));
+
+                    let batch_reader =
+                        arrow_reader.get_record_reader(records.try_into().unwrap())?;
+                    for batch in batch_reader {
+                        let batch = batch?;
+                        writer.write_partition(batch, partition_values).await?;
+                    }
+
+                    actions.push(self.create_remove(path, partition_values, size)?);
+
+                    self.metrics.num_files_removed += 1;
+                    self.metrics.files_removed.total_files += 1;
+                    self.metrics.files_removed.total_size += size;
+                    self.metrics.files_removed.max =
+                        std::cmp::max(self.metrics.files_removed.max, size);
+                    self.metrics.files_removed.min =
+                        std::cmp::min(self.metrics.files_removed.min, size);
+                }
+
+                //Save the file to storage and create corresponding add and delete actions. Do not commit yet.
+                let add_actions = writer.flush().await?;
+                for mut add in add_actions {
```

Comment:
In theory, we should be only getting one add action here, correct? Wondering if we've checked that.